chisel/src/options.rs

```diff
@@ -0,0 +1,127 @@
+//! MODULE_OPTIONS: A list of options set for the modules being invoked in oneliner mode.
+//! FILE: Sets the input file path in oneliner mode.
+//! OUTPUT_PATH: Sets the path to write any mutated binaries in oneliner mode.
+//! OUTPUT_MODE: Sets the format in which to output mutated binaries.
+//!      - wasm: default binary mode. disallowed when writing to stdout.
+//!      - hex: write the output in hex. recommended if writing to stdout.
+//!      - wat: write the output in disassembled (.wat) format.
+
+use std::collections::HashMap;
+use std::ops::Deref;
+
+use clap::ArgMatches;
+
+/// Key-value structure for immutable CLI options. Used for storing utility options and
+/// configurations in oneliner mode.
+pub struct ChiselFlags(HashMap<String, String>);
+
+impl ChiselFlags {
+    /// Sets the value of `key`.
+    pub fn set(&mut self, key: &str, value: &str) {
+        self.0.insert(key.to_string(), value.to_string());
+    }
+
+    /// Gets the value of `key`.
+    pub fn value_of(&self, key: &str) -> Option<&str> {
+        match self.0.get(key) {
+            Some(s) => Some(s.as_str()),
+            None => None,
+        }
+    }
+
+    /// Compares the value of `key` to the passed `val`.
+    pub fn value_eq(&self, key: &str, val: &str) -> bool {
+        match self.value_of(key) {
+            Some(s) => s == val,
+            None => false,
+        }
+    }
+
+    /// Apply all flags passed from CLI
+    pub fn apply(&mut self, matches: &ArgMatches) {
+        if matches.is_present("NO_RECOVER") {
+            self.set("util.norecover", "true");
+        }
+        if matches.is_present("VERBOSE") {
+            self.set("util.debugging", "true");
+        }
+        if let Some(value) = matches.value_of("CONFIG") {
+            self.set("run.config.path", value);
+        }
+        if let Some(values) = matches.values_of("MODULES") {
+            let values_collected = values.fold(String::new(), |mut acc, val| {
+                acc.push_str(&format!("{},", val));
+                acc
+            });
+            self.set("oneliner.modules", &values_collected);
+        }
+        if let Some(values) = matches.values_of("MODULE_OPTIONS") {
+            let values_collected = values.fold(String::new(), |mut acc, val| {
+                acc.push_str(&format!("{},", val));
+                acc
+            });
+            self.set("oneliner.modules.options", &values_collected);
+        }
+        if let Some(value) = matches.value_of("FILE") {
+            self.set("oneliner.file", value);
+        }
+        if let Some(value) = matches.value_of("OUTPUT_PATH") {
+            self.set("oneliner.output", value);
+        }
+        if let Some(value) = matches.value_of("OUTPUT_MODE") {
+            match value {
+                val @ "bin" | val @ "wat" | val @ "hex" => {
+                    self.set("output.mode", val);
+                }
+                _ => panic!("CLI parser only accepts 'bin', 'wat', or 'hex'"),
```

Comment:
Why not return an error here?

Reply:
This function does not return a set of flags, it applies arguments from `clap` to the flag set.

Reply:
Terrible. Is there a way to signal to clap what values are accepted?

Reply:
It does: https://docs.rs/clap/2.33.0/clap/struct.Arg.html#method.possible_values

Reply:
Yes, in the app declaration. It should be done there, which is why it makes sense to panic here. This simply should be unreachable.

Reply:
This panic is fine as long as the above `possible_values` will be used so that clap doesn't even let this pass through to the panic.

Reply:
exactly